File: Tree/LeetCode/Binary_Tree_Preorder_Traversal.go
package leetCode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}
type Stack struct {
	items []*TreeNode
}

func preorderTraversal(root *TreeNode) []int {
	var stk Stack
	res := []int{}
	for root != nil || !stk.IsEmpty() {
		for root != nil {
			res = append(res, root.Val)
			stk.Push(root)
			root = root.Left
		}
		root = stk.Pop()

		root = root.Right
	}
	return res
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}
func (s *Stack) Push(t *TreeNode) {
	s.items = append(s.items, t)
}
func (s *Stack) Pop() *TreeNode {
	if len(s.items) == 0 {
		return nil
	}
	res := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return res
}

File: Tree/LeetCode/Binary_Tree_Preorder_Traversal.cs
using System.Collections.Generic;
using System.Linq;

namespace Tree.LeetCode
{
    public class Binary_Tree_Preorder_Traversal
    {
        public IList<int> PreorderTraversal(TreeNode root)
        {
            var res = new List<int>();
            if (root == null)
            {
                return res;
            }
            if (root.left == null && root.right == null)
            {
                res.Add(root.val);
                return res;
            }

            Stack<TreeNode> stk = new Stack<TreeNode>();
            stk.Push(root);
            while (stk.Any())
            {
                var item = stk.Pop();
                res.Add(item.val);
                if (item.right != null)
                {
                    stk.Push(item.right);
                }
                if (item.left != null)
                {
                    stk.Push(item.left);
                }
            }
            return res;
        }
    }


}